Python/errors.c

```diff
@@ -151,7 +151,14 @@ _PyErr_SetObject(PyThreadState *tstate, PyObject *exception, PyObject *value)
     }
     Py_XINCREF(value);
     /* Normalize the exception */
-    if (value == NULL || (PyObject *)Py_TYPE(value) != exception) {
+    int is_subclass = 0;
+    if (value != NULL) {
+        is_subclass = PyObject_IsSubclass((PyObject*)Py_TYPE(value), exception);
```

Comment:
IIUC `PyObject_IsSubclass` is relatively expensive. What if we skip calling it when `(PyObject *)Py_TYPE(value) != exception`?

Reply:
That's not equivalent though.

Reply:
I think Guido meant `is_subclass = (Py_TYPE(value) == exception || PyObject_IsSubclass(Py_TYPE(value), exception)`.

Reply:
You'd have to do
```
if ((PyObject *)Py_TYPE(value) != exception)
    is_subclass = 1;
else
    is_subclass = PyObject_IsSubclass((PyObject*)Py_TYPE(value), exception);
```
(After the NULL check.)

Don't hold up the alpha for this though.

Reply:
(Whoops, note `==`, not `!=`. Corrected in place.)

Reply:
Do you mean when it's not an exception, as in `!PyExceptionInstance_Check(value)`?

Reply:
Never mind me.

Reply:
After checking PyObject_IsSubclass, I don't think it's worth it: after a bit of indirection through tstate it does:

```
    /* We know what type's __subclasscheck__ does. */
    if (PyType_CheckExact(cls)) {
        /* Quick test for an exact match */
        if (derived == cls)
            return 1;
```
... so it's only an extra pointer comparison and a couple of quick C function calls more expensive in this case.